rask-engine/src/resources/library.rs

```diff
@@ -0,0 +1,73 @@
+use super::Resource;
+use crate::EngineError;
+
+/// The library is used to store and retrieve resources.
+pub struct ResourceTable(&'static mut [Resource]);
+
+macro_rules! get_store {
+    ($type: ty, $enum_type: ident) => {
+        impl GetStore<$type> for ResourceTable {
+            unsafe fn get(&'static self, id: usize) -> Result<&'static $type, EngineError> {
+                self.index_check(id)?;
+                match &self.0[id] {
+                    Resource::$enum_type(value) => Ok(&value),
+                    _ => Err("Wrong resource type".into()),
+                }
+            }
+            unsafe fn store(&mut self, data: $type, id: usize) -> Result<(), EngineError> {
+                self.index_check(id)?;
+                Ok(self.0[id] = Resource::$enum_type(data))
+            }
+        }
+    };
+}
+
+pub trait GetStore<T> {
+    /// Retrieve a resource from the library.
+    ///
+    /// # Safety
+    ///
+    /// The function is not thread safe.
+    unsafe fn get(&'static self, id: usize) -> Result<&'static T, EngineError>;
+    /// Store a resource to the library
+    ///
+    /// # Safety
+    ///
+    /// The function is not thread safe.
+    unsafe fn store(&mut self, data: T, id: usize) -> Result<(), EngineError>;
+}
+
+impl ResourceTable {
+    /// Create a new library at a specific position in memory.
+    ///
+    /// # Safety
+    ///
+    /// The function is safe as long as the memory from memory_offset to memory_offset + CATALOG_SIZE * sizeof(Resource)
+    pub unsafe fn new(memory_offset: usize, catalog_size: usize) -> Self {
+        ResourceTable(core::slice::from_raw_parts_mut(
+            memory_offset as *mut Resource,
+            catalog_size,
+        ))
+    }
+
+    pub unsafe fn init(&mut self) {
```

Comment:
It has to be possible to have an uninitialized table, that's the whole point.
otherwise the graphics thread  would override entries made on the logic thread
I think is is technically not even needed, since wasm memory in initialized with 0,  this function is just for added safety.
coult rename to clear()

Reply:
Yes, I see. I think, `new` should construct a new `ResourceTable` (corresponding to `new` and then `init`), and than there can be a `from_address`-function (corresponding to just `new`)